### ISSUE TITLE
[Linux] Incorrect OpenGL usage results in sRGB textures not actually loading

### ISSUE THREAD

**Issue by iamgreaser** [Yellow-Dog-Man/Resonite-Issues#69]
### Describe the bug?

sRGB textures do not load due to an invalid `format` argument in a call to `glTexSubImage2D()`. For reference, I am talking  about the `format` parameter for this function prototype:

```
void glTexSubImage2D(GLenum target, GLint level, GLint xoffset, GLint yoffset, GLsizei width, GLsizei height, GLenum format, GLenum type, const void *pixels);
```

Any case where the `format` field is set to `GL_SRGB_ALPHA`, set it to `GL_RGBA` instead. Do a find and replace through your codebase. You're welcome. (If you're stuck, I have an `LD_PRELOAD` library which works around this bug.)

This is ultimately caused by the fact that NVidia's proprietary OpenGL driver is so full of workarounds for bugs that the driver itself *is* a bug, and relying on it results in everything being wrong. Seriously, for the love of everything that is worth loving, use literally any other driver when testing OpenGL stuff. (For maximum speed, the zink driver paired with NVidia's proprietary Vulkan driver might work.)

### To Reproduce

Open the Linux version with literally any driver that isn't NVidia's proprietary OpenGL driver.

If the floor texture of your local home looks screwed up and/or your dash looks screwed up and/or potentially blank, congratulations, you've reproduced the issue.

If you have an integrated GPU, use that. You may want to lower your resolution.

Otherwise, use llvmpipe (`LIBGL_ALWAYS_SOFTWARE=yes`) and lower your resolution to something like 640x360 to make the framerate close to bearable.

If you know what you're doing, you may have some success with the zink driver paired with NVidia's proprietary Vulkan driver.

### Expected behavior

Textures should actually load their data in.

### Resonite Version Number

2023.10.13.743

### What Platforms does this occur on?

Linux

### Additional Context

I've told Geenz and maybe Frooxius about this, so this issue is mostly to ensure that it is tracked properly. I'm filing this issue because I've seen a screenshot in another issue where the screwed up floor is clearly visible.

### Reporters

GreaseMonkey

**Comment by yoshiyoshyosh** [comment]
> If you have an integrated GPU, use that. You may want to lower your resolution.

I observe this behavior on my Integrated Graphics CPU (AMD 5600G).

> Otherwise, use llvmpipe (`LIBGL_ALWAYS_SOFTWARE=yes`) and lower your resolution to something like 640x360 to make the framerate close to bearable.

While this does seemingly fix the glitched texture loading for me, it does *not* seemingly fix the blank/screwed up dashboard for me (as reported here #63)

**Comment by iamgreaser** [comment]
Huh, you must've posted your report just after I started writing mine.

I have an LD_PRELOAD library to work around this actually. I've attached it here, CC0, comes with source code and a precompiled version. Try it and see if you still have the issue.

[bug-69-workaround.zip](https://github.com/Yellow-Dog-Man/Resonite-Issues/files/12900246/bug-69-workaround.zip)

**Comment by yoshiyoshyosh** [comment]
Can confirm the `LD_PRELOAD` fixes the issue.

**Comment by rcelyte** [comment]
> Do a find and replace through your codebase

More specifically, the misbehaving code lies within `libNativeGraphics.so`

**Comment by rcelyte** [comment]
Here's a modloader-based fix for people who don't want to mess around with LD_PRELOAD: https://github.com/rcelyte/ResoniteTextureFix

**Comment by rcelyte** [comment]
To detail exactly what needs to be changed: The line in `UnityFrooxEngineRunner.Helper.ToOpenGL()` under `case Elements.Assets.TextureFormat.BGRA32:` with `.baseFormat = OpenGL.TextureFormat.GL_SRGB_ALPHA` needs to be changed to use `OpenGL.TextureFormat.GL_RGBA`.

**Comment by ZarathustraDK** [comment]
Gonna risk my neck here :) @Frooxius , we could reaaally use rcelyte's fix above if possible. it's basically the only thing blocking Resonite from running flawlessly natively on Linux on the Monado/Opencomposite-stack afaik.

**Comment by Frooxius** [comment]
I could take this one, but this is currently assigned to @Geenz

**Comment by ProbablePrime** [comment]
I'm seeing inconsistent reporting of GL_SRGB_ALPHA.

For example it is mentioned here: https://registry.khronos.org/OpenGL-Refpages/gl4/html/glTexImage2D.xhtml
But not here? https://registry.khronos.org/OpenGL-Refpages/gl4/html/glTexSubImage2D.xhtml#parameters as you said.
Its also here: https://docs.gl/gl3/glTexImage2D
But that's the same content as the Khronos link just formatted differently.

A particular source for "only works on certain non-conformant drivers which silently replace it with GL_RGBA internally." out in the wild would help a lot too. 

Happy to make the change, I just want to leave a comment behind talking about why and get a lot happier making a change when I understand why.

**Comment by iamgreaser** [comment]
> I'm seeing inconsistent reporting of `GL_SRGB_ALPHA`.

It's consistent. A call to `glTexImage2D` accepts an `internalFormat` argument, while a call to `glTexSubImage2D` does not. The bug is caused by Resonite is filling in the `format` argument with an invalid value, which is a parameter of both functions.

> A particular source for "only works on certain non-conformant drivers which silently replace it with GL_RGBA internally." out in the wild would help a lot too.

As in, "I have heard that the current code apparently works on NVidia's proprietary driver, while my personal experience with other drivers has been that they correctly fail in this case according to the specification". Which says, in the case of `glTexSubImage2D`:

> `GL_INVALID_ENUM` is generated if `format` is not an accepted format constant.

But it would be better if someone using NVidia's propietary driver on Linux were to test one of the mods which fixes the problem (be it my `LD_PRELOAD`-based one or one of the modloader-based ones - I thought there was at least one more floating around but I could be wrong).

If you've ever done enough development for OpenGL to have to deal with different drivers screwing things up, either you're using a non-NVidia driver in which case your experience is "some drivers need the occasional workaround here and there", or you're using the official NVidia driver in which case your experience is "why do I need to do so many workarounds for every other driver" and/or "huh, I really do need to read the spec to get things working properly".

Anyway, this critical issue has been left in a "here's roughly how you fix it, please do it already" state for about 9 months now, and by "critical" I mean "the Linux client is unusable without it for most people". It's not the only critical problem I know of (the Brotli one also needs to be fixed in order to be able to save anything), but it still needs to be fixed.

I don't see how stalling this issue any further helps anyone. But fixing it would definitely help with my rather drained morale.

rcelyte has explained exactly what needs to be done.

**Comment by ProbablePrime** [comment]
I don't appreciate the implication that I'm stalling. Its 2am on a Wednesday and instead of sleeping I'm looking into it. By my opinion any action here is welcome from the radio silence we've been giving here on GH to this issue.

Generally, I refuse to commit code I don't understand. It's a very bad idea. Every commit I make has to be from a place of knowledge and understanding and not from a place of someone on the internet pressuring me to "just do it".

Please ensure you're following our Code of Conduct and reading our "How we prioritize" document which very clearly talks about all of this.

**Comment by iamgreaser** [comment]
Please understand that some of us are fed up with the constant apparent pattern of "it pertains to the Linux client therefore we're not going to bother with it". But anything to help fix that is a welcome change from that. Sorry if I kinda dumped some of my frustration on you.

Enjoy your sleep.

**Comment by ProbablePrime** [comment]
That isn't the reason. The reason is time as evidenced by me trying to do this in my off time rather than taking a break and not working. As I've said numerous times in relation to linux over the past week. I want to fix it. I just need to find the time.

Please ensure you're reading: https://github.com/Yellow-Dog-Man/Resonite-Issues/blob/main/HOW_WE_PRIORITIZE.md to understand more about why Linux issues might not be being prioritized instead of assuming various things about me and the development team.

**Comment by BlueCyro** [comment]
This isn't super applicable to the issue, but I do want to chime in and say that Resonite isn't 'unusable' as is suggested. You can very well run it through proton and play it that way with just about the same featureset/performance. Sure it's not native, but I feel saying Resonite is unusable on linux is an extremely hyperbolic view of the situation here.

Having the native build work right would be nice, but the game runs just fine through proton, and it really doesn't matter which way you run it in the end - you get just about the same experience.

**Comment by ljoonal** [comment]
> You can very well run it through proton...

Not a given on all Linux setups, proton has more issues on my setup than the native version. <sub>And actually fixing the proton issues is harder than the Linux ones, as those tend to be tied to window manager implementations, environment vars, wine version, and other things which change a lot between each OS install, so there's way less discussion on thowe issues even existing in the first place.</sub>

I fully agree with Prime though on:
> Generally, I refuse to commit code I don't understand. It's a very bad idea.



IMO this issue could also be aided by the community documenting all the required installation steps and hacks on the wiki, and then having that linked from the more visible & official Linux chats&channels. Instead of the current state, where just to get Resonite to be sort of usable on my machines, I need to dig trough countless random GH issues and re-figure out which and how the patches I need can be created every time I fresh install Resonite.

**Comment by rcelyte** [comment]
> I'm seeing inconsistent reporting of GL_SRGB_ALPHA.
> 
> For example it is mentioned here: https://registry.khronos.org/OpenGL-Refpages/gl4/html/glTexImage2D.xhtml But not here? https://registry.khronos.org/OpenGL-Refpages/gl4/html/glTexSubImage2D.xhtml#parameters as you said. Its also here: https://docs.gl/gl3/glTexImage2D But that's the same content as the Khronos link just formatted differently.
> 
> A particular source for "only works on certain non-conformant drivers which silently replace it with GL_RGBA internally." out in the wild would help a lot too.
> 
> Happy to make the change, I just want to leave a comment behind talking about why and get a lot happier making a change when I understand why.

To explain the OpenGL-isms, a texture handle encompasses two distinct resources driver-side: the backing image data and the sampler used to interpret that data within shaders. With `glTexImage2D()`, somewhat unintuitively, the `internalformat` parameter specifies how the sampler interprets data while a combination of the `format` and `type` parameters defines the actual backing data layout (at least, that's the best way I can describe it). `glTexSubImage2D()` does a plain byte transfer without modifying the sampler or allocation state, so the concept of an `internalformat` doesn't exist for it.

For the "only works on certain non-conformant drivers" bit, these kinds of odd quirks and differences between drivers aren't usually documented anywhere, but you can confirm for yourself by running Resonite over Mesa's hardware agnostic Zink (GL on Vulkan) or LLVMPipe (GL on CPU) drivers. As far as anyone's aware, every driver besides Nvidia's proprietary driver follows the spec and treats `GL_SRGB_ALPHA` as an invalid enum for the `format` parameter of `glTexImage*D()` and `glTexSubImage*D()`.

**Comment by regalialong** [comment]
> But it would be better if someone using NVidia's proprietary driver on Linux were to test one of the mods which fixes the problem (be it my LD_PRELOAD-based one or one of the modloader-based ones - I thought there was at least one more floating around but I could be wrong).

Hey @iamgreaser, figured I'd chime in since there was some confusion at LVRA about this.
This workaround is also needed on the proprietary driver 555.52.04 for me.

Without the preload, my menu does not render (notice the black line around the loading)

![image](https://github.com/Yellow-Dog-Man/Resonite-Issues/assets/126194895/a76f055d-b62f-4ec2-9094-cbbc12ee63c2)

With the preload, menu actually does render and the circle gains color
![image](https://github.com/Yellow-Dog-Man/Resonite-Issues/assets/126194895/a54dd4f0-624c-45a9-ae3f-aa709fc71899)
![image](https://github.com/Yellow-Dog-Man/Resonite-Issues/assets/126194895/0f988e17-57f2-47d6-af09-05059012cf9a)

Not sure if the driver became compliant in the time this was reported, but assuming that it was everyone but Nvidia and now Nvidia is affected, this is a generic showstopper for the Linux client.

**Comment by Geenz** [comment]
I've pushed a tentative fix into a PR for this.

**Comment by Frooxius** [comment]
Merged fix in 2024.6.24.1317, thanks!
